records/2024-06-04.manifests-kustomize-cicd.md

```diff
@@ -0,0 +1,140 @@
+
+```shell
+kubectl set image deployment.apps/api api=$DOCKER_SLUG:$DOCKER_TAG -n=notification-canada-ca
+```
+
+##### Diagram
+
+![Staging Current CICD](./diagrams/2024-06-04.manifests-current-cicd/staging-cicd.png)
+
+#### Global Manifests Rollouts
+
+When a global manifests rollout is performed by the merge to main (staging) github workflow, all of the Kustomize code is deployed.
+
+```shell
+notification-manifests/env/staging kubectl apply -k .
+```
+
+### Production
+
+#### Individual Application Rollouts
+
+This concept does not exist for production. We only rollout all of Kustomize.
+
+#### Global Manifests Rollouts
+
+Every night, the Notify PR Bot reads the latest github commit sha's for each repository, and if they've changed, creates a pull request that changes the image tag values in the Kustomization.yaml for production. Once this is merged, all of the kustomize code is deployed at once.
+
+Since the PR bot does not touch the helmfile side of things, we also do a single global apply on helmfile with "latest" tags.
+
+#### Diagram
+
+![Production Current CICD](./diagrams/2024-06-04.manifests-current-cicd/production-cicd.png)
+
+### Problems
+
+#### Two Different Deployment Technologies
+
+We are currently managing two different Kubernetes deployment methods with Kustomize and Helmfile. It would be beneficial to have everything under one technology.
+
+#### Staging Releases
+
+Because we do not actually update any manifests code when doing an individual application rollout, we lose all context of which version we should be using when we do a global manifest rollout with merge to main (staging). To overcome this, the staging Kustomize code is set to always use the "latest" tag. This means when we apply kustomize the explicitly versioned docker tag is overwritten with "latest". This works because the docker build and push job for each application tags as both latest and the github sha. It is however not ideal since it becomes difficult to track versions in staging, and also creates a delta between how staging and production releases are handled through release pipelines.
+
+This issue is a problem regardless of whether we use Kustomize or Helmfile.
+
+#### Helmfile Image Versions
+
+For helmfile releases, we are relying on the helmfile logic to set the image versions in the overrides. For example, we may have something like the code block below, which hard codes the tag for production, but accepts either an environment variable or will default to "latest" if it does not exist.
+
+```yaml
+      {{ if eq .Environment.Name "production" }}
+      image:
+        tag: release1.24
+      {{ else }}
+      image:
+        tag: {{ env "DOCKER_TAG" | default "latest" }}
+      {{ end }}
+```
+
+This is an ok solution, but it's not ideal to have to modify the individual override files to set image versions for production manually. It would be ideal to have this done automatically like with the PR bot.
```

Comment:
I agree with keeping pr-bot managing this and I think this should be a requirement toward using Helm.